### PR TITLE
[feature] #2000: Disallow empty names

### PR DESCRIPTION
**Pull request by ales-tsurko**
This PR disallows to use empty strings for `iroha_data_model::Name`. Thus, it forbids empty `AccountId`, `DomainId` etc. In case of `Name`'s initialization from an empty string, an error is raised. For discussion see the linked issue (#2000).

### CODE REVIEW

File: data_model/src/name.rs [hunk 1]
```diff
@@ -22,9 +22,9 @@ use crate::{ParseError, ValidationError};
 pub struct Name(ConstString);
 
 impl Name {
-    pub(crate) const fn empty() -> Self {
-        Self(ConstString::new())
-    }
+    // pub(crate) const fn empty() -> Self {
```
Comment (Erigara) [inline data_model/src/name.rs:25]: commented code

File: data_model/src/name.rs [hunk 2]
```diff
@@ -80,12 +81,7 @@ impl FromStr for Name {
     type Err = ParseError;
 
     fn from_str(candidate: &str) -> Result<Self, Self::Err> {
-        Self::validate_str(candidate).map(|_| {
-            if candidate.is_empty() {
-                return Self::empty();
-            }
-            Self(ConstString::from(candidate))
-        })
+        Self::validate_str(candidate).map(|_| Self(ConstString::from(candidate)))
```
Comment (Erigara) [inline data_model/src/name.rs:84]: looks much cleaner now without `is_empty` branch!

**Review by Erigara — CHANGES_REQUESTED** [review]
It looks good in general, however I think we should remove the commented out code and also add a test case to check that trying to pass an empty string returns an error.

File: data_model/src/account.rs [hunk 3]
```diff
@@ -461,7 +454,9 @@ impl FromStr for Id {
 
     fn from_str(string: &str) -> Result<Self, Self::Err> {
         if string.is_empty() {
-            return Ok(Self::empty());
+            return Err(ParseError {
+                reason: "`AccountId` can not be empty",
```
Comment (outoftardis) [inline data_model/src/account.rs:458]:
```suggestion
                reason: "`AccountId` cannot be empty",
```

File: data_model/src/asset.rs [hunk 4]
```diff
@@ -654,7 +647,9 @@ impl FromStr for DefinitionId {
 
     fn from_str(string: &str) -> Result<Self, Self::Err> {
         if string.is_empty() {
-            return Ok(Self::empty());
+            return Err(ParseError {
+                reason: "`DefinitionId` can not be empty",
```
Comment (outoftardis) [inline data_model/src/asset.rs:651]:
```suggestion
                reason: "`DefinitionId` cannot be empty",
```

File: data_model/src/name.rs [hunk 5]
```diff
@@ -52,9 +52,10 @@ impl Name {
     /// Fails if not valid [`Name`].
     fn validate_str(candidate: &str) -> Result<(), ParseError> {
         if candidate.is_empty() {
-            return Ok(());
+            return Err(ParseError {
+                reason: "`Name` can not be empty",
```
Comment (outoftardis) [inline data_model/src/name.rs:56]:
```suggestion
                reason: "`Name` cannot be empty",
```

**Review by outoftardis — COMMENTED** [review]
`cannot` is just far more common compared to `can not`, so I suggest this change

File: data_model/src/name.rs (hunk 1 above)
Comment (ales-tsurko) [reply data_model/src/name.rs]: Ah, I just forgot to push it.

**Review by outoftardis — APPROVED** [review]

File: data_model/src/account.rs [hunk 6]
```diff
@@ -461,7 +454,9 @@ impl FromStr for Id {
 
     fn from_str(string: &str) -> Result<Self, Self::Err> {
         if string.is_empty() {
```
Comment (Erigara) [inline data_model/src/account.rs:456]: I think this check is unnecessary because the next check for the correct format will always fail with an empty string, but it's useful in terms of error readability. 
Should we keep it? @ales-tsurko what do you think?

Comment (ales-tsurko) [reply data_model/src/account.rs:456]: I think this check is unnecessary because the next check for the correct format will always fail with an empty string, but it's useful in terms of error readability.
Should we keep it? @ales-tsurko what do you think?

It has different `reason` description. The first is more specific and related to particular case (i.e. empty string).

Comment (Erigara) [reply data_model/src/account.rs:456]: ok, resolve then

**Review by Erigara — APPROVED** [review]
